src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionSelector.cs

```diff
@@ -234,5 +283,177 @@ protected virtual IReadOnlyList<ActionDescriptor> SelectBestActions(IReadOnlyLis
                 return EvaluateActionConstraintsCore(context, actionsWithoutConstraint, order);
             }
         }
+
+        // The action selector cache stores a mapping of route-values -> action descriptors for each 'known' set of
+        // of route-values. We actually build two of these mappings, one for case-sensitive (fast path) and one for
+        // case-insensitive (slow path).
+        //
+        // This is necessary because MVC routing/action-selection is always case-insensitive. So we're going to build
+        // a case-sensitive dictionary that will behave like the a case-insensitive dictionary when you hit one of the
+        // canonical entries. When you don't hit a case-sensitive match it will try the case-insensitive dictionary
+        // so you still get correct bheaviors.
+        //
+        // The difference here is because while MVC is case-insensitive, doing a case-sensitive comparison is much 
+        // faster. We also expect that most of the URLs we process are canonically-cased because they were generated
+        // by Url.Action or another routing api.
+        //
+        // This means that for a set of actions like:
+        //      { controller = "Home", action = "Index" } -> HomeController::Index1()
+        //      { controller = "Home", action = "index" } -> HomeController::Index2()
```

Comment:
Action constraints can disambiguate. If you think about this, any case that explains why this behavior is necessary will require action constraints to disambiguate.